File: geohash-ii/geohash-ii.py
class GeoHash:
    # @param {string} geohash a base32 string
    # @return {double[]} latitude and longitude a location coordinate pair
    def decode(self, geohash):
        # Write your code here
        base32 = "0123456789bcdefghjkmnpqrstuvwxyz"
        n = len(geohash)
        b = []
        for i in xrange(n):
            c = geohash[i]
            for j in xrange(32):
                if c == base32[j]:
                    tb = []
                    for k in xrange(5):
                        tb.append(j & 1)
                        j >>= 1
                    for tbit in tb[::-1]:
                        b.append(tbit)
                    break
        m = len(b)
        lo = self.get_db(b[::2], -180, 180, m - m / 2)
        la = self.get_db(b[1::2], -90, 90, m / 2)
        return [la, lo]
    
    def get_db(self, b, l, r, n):
        for i in xrange(n):
            m = (l + r) * 0.5
            if b[i] == 0:
                r = m
            else:
                l = m
        return (l + r) * 0.5

File: geohash-ii/geohash-ii.cc
class GeoHash {
    public:
    /**
     * @param geohash a base32 string
     * @return latitude and longitude a location coordinate pair
     */
    vector<double> decode(string& geohash) {
        // Write your code here
        int n = geohash.length();
        for (int i = 0; i < n; ++i) {
            get_bin(geohash[i], i * 5);
        }
        int c_la, c_lo;
        c_la = n * 5 / 2;
        c_lo = n * 5 - c_la;
        
        for (int i = 0; i < c_la; ++i) {
            b_lo[i] = b[i << 1];
            b_la[i] = b[i << 1 | 1];
        }
        if (c_lo > c_la) {
            b_lo[c_la] = b[c_la << 1];
        }
        double la, lo;
        lo = get_db(b_lo, -180, 180, c_lo);
        la = get_db(b_la, -90, 90, c_la);
        return vector<double>{la, lo};
    }
    
    void get_bin(char c, int offset) {
        for (int i = 0; i < 32; ++i) {
            if (base32[i] == c) {
                for (int j = offset + 4; j >= offset; --j) {
                    b[j] = i & 1;
                    i >>= 1;
                }
                return;
            }
        }
    }
    
    double get_db(int *b, double l, double r, int n) {
        double m;
        for (int i = 0; i < n; ++i) {
            m = (l + r) / 2;
            if (b[i]) {
                l = m;
            } else {
                r = m;
            }
        }
        return (l + r) / 2; // imprecise
    }
    
    private:
    string base32 = "0123456789bcdefghjkmnpqrstuvwxyz";
    int b[60];
    int b_la[30], b_lo[30];
};
